src/SelfGravity/CPU_Poisson/CPU_ComputeGREP.cpp

```diff
@@ -41,59 +44,77 @@ void CPU_ComputeGREP( const int lv, const double Time, const Profile_t *DensAve,
    double *Vr        = VrAve  ->Data;
    double *Pres      = PresAve->Data;
 
+   double *Weight       = new double [NBin];  // volume                           of each bin
+   double *Edge_L       = new double [NBin];  // left edge                        of each bin
+   double *Vr_L         = new double [NBin];  // radial velocity at the left edge of each bin
    double *Mass_NW      = new double [NBin];  // enclosed Newtonian mass for \bar_Phi(r)     in Eq. (7)
-   double *Mass_TOV     = new double [NBin];  // enclosed TOV mass       for \bar_Phi(r)_TOV in Eq. (7)
-   double *Mass_TOV_USG = new double [NBin];  // enclosed TOV mass in the previous iteration
+   double *Mass_TOV     = new double [NBin];  // enclosed TOV       mass for \bar_Phi(r)_TOV in Eq. (7)
+   double *Mass_TOV_USG = new double [NBin];  // enclosed TOV       mass in the previous iteration
    double *Dens_TOV     = new double [NBin];  // empirical TOV density                          Eq. (4)
    double *Gamma_TOV    = new double [NBin];  // metric function                                Eq. (5)
-   double *dVol         = new double [NBin];  // volume    of each bin
-   double *EdgeL        = new double [NBin];  // left edge of each bin
 
-   for (int i=0; i<NBin;   i++)   Mass_TOV_USG[i] = 0.0;
-   for (int i=0; i<NBin;   i++)   Gamma_TOV   [i] = 1.0;
+// (1) preparation stage
+   for (int i=0; i<NBin; i++)   Mass_TOV_USG[i] = 0.0;
+   for (int i=0; i<NBin; i++)   Gamma_TOV   [i] = 1.0;
+
+                                Edge_L[0] = 0.0;
+   for (int i=1; i<NBin; i++)   Edge_L[i] = ( GREP_LOGBIN ) ? sqrt( Radius[i] * Radius[i-1] )
+                                                            : 0.5*( Radius[i] + Radius[i-1] );
+
+// find the maximum radius that the stored weights are reliable
+   for (int d=0; d<3; d++)
+   {
+      MaxRadius = MIN(  MIN( amr->BoxSize[d] - DensAve->Center[d], DensAve->Center[d] ),
+                        MaxRadius  );
+   }
 
-                                  EdgeL       [0] = 0.0;
-   for (int i=1; i<NBin;   i++)   EdgeL       [i] = ( GREP_LOGBIN ) ? sqrt( Radius[i-1] * Radius[i] )
-                                                                    : 0.5*( Radius[i-1] + Radius[i] );
+// update the weight
+   for (int i=0; i<NBin-1; i++)
+   {
+      Weight[i] = ( Edge_L[i+1] > MaxRadius )
+                ? FourThirdPI * ( CUBE(Edge_L[i+1]) - CUBE(Edge_L[i]) )
+                : DensAve->Weight[i];
+   }
 
-   for (int i=0; i<NBin-1; i++)   dVol        [i] = FourThirdPI * ( CUBE( EdgeL[i+1] ) - CUBE( EdgeL[i]      ) );
-                                  dVol   [NBin-1] = FourThirdPI * ( CUBE( MaxRadius  ) - CUBE( EdgeL[NBin-1] ) );
+   Weight[NBin-1] = FourThirdPI * ( CUBE(DensAve->MaxRadius) - CUBE(Edge_L[NBin-1]) );
```

Comment:
Could you remind me why we don’t simply use `DensAve->Weight[i]` for all bins, including the last bin (`NBin-1`)? I thought this approach was necessary to ensure mass conservation. Or is this functionality implemented in a different branch?

If I recall correctly, we discussed this a while ago. Using `DensAve->Weight[i]` also results in more consistent mass profiles on different AMR levels, particularly at lower levels where spatial resolution is poor.

Reply:
This approach ensures consistency in enclosed mass compared to other codes. Imagine a cube (simulation box) within a sphere (GREP domain). The six hemispheres will not be counted when using the weight, as they are not covered by the simulation domain, leading to an underestimated enclosed mass.

Reply:
Thanks for the elaboration. I now recall the reasoning, and I agree with it. Could you also add this explanation as a comment in the code for future reference?